src/utils/trpc.ts

```diff
@@ -9,13 +14,65 @@ import { TRPCWithErrorCodeSchema } from '~/utils/error'
 import type { AppRouter } from '~/server/modules/_app'
 import { getBaseUrl } from './getBaseUrl'
 import { type TRPC_ERROR_CODE_KEY } from '@trpc/server/rpc'
+import { LOGGED_IN_KEY } from '~/constants/insecureCookies'
+import { SIGN_IN } from '~/lib/routes'
+import { CALLBACK_URL_KEY } from '~/constants/params'
+import { deleteCookie } from 'cookies-next'
+import { observable } from '@trpc/server/observable'
 
 const NON_RETRYABLE_ERROR_CODES: Set<TRPC_ERROR_CODE_KEY> = new Set([
   'UNAUTHORIZED',
   'FORBIDDEN',
   'NOT_FOUND',
 ])
 
+export const custom401Link: TRPCLink<AppRouter> = () => {
+  // here we just got initialized in the app - this happens once per app
+  // useful for storing cache for instance
+  return ({ next, op }) => {
+    // this is when passing the result to the next link
+    // each link needs to return an observable which propagates results
+    return observable((observer) => {
+      const unsubscribe = next(op).subscribe({
+        next(value) {
+          observer.next(value)
+        },
+        // Handle 401 errors
+        error(err) {
```

Comment:
Done and actually improved by using localStorage (again), allowing window to dispatch an event, allowing react to rerender instead of bubbling the error